### PR TITLE
Event ordering preparation

### PR DESCRIPTION
**Pull request by fguimond**
Signed-off-by: Francis Guimond <francis@sensu.io>

## What is this change?

In preparation for Postgres ordering/filtering/pagination:

* Add sort fields to event api
* Add offset and ordering to SelectionPredicate

## Why is this change necessary?

In preparation for Postgres ordering/filtering/pagination.

## Does your change need a Changelog entry?

No.

## Do you need clarification on anything?

No.

## Were there any complications while making this change?

No.

## Have you reviewed and updated the documentation for this change? Is new documentation required?

not required.

## How did you verify this change?

Unit tests.

## Is this change a patch?

No.

### CODE REVIEW

File: backend/store/store.go [hunk 1]
```diff
@@ -99,14 +99,18 @@ type SelectionPredicate struct {
 	// returned empty from the store, it indicates that there's no additional
 	// resources available
 	Continue string
+	// Offset indicates the index of the first resource to retrieve
+	Offset int64
```
Comment (jamesdphillips) [inline backend/store/store.go:103]: Did we decide if we wanted to do this or not?

File: api/core/v2/event_sort.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	case EventSortLastOk:
+		sortIf = EventsByLastOk(events, asc)
+	case EventSortSeverity:
+		sortIf = EventsBySeverity(events, asc)
+	default:
+		sortIf = EventsByTimestamp(events, asc)
+	}
+
+	sort.Sort(sortIf)
+}
```
Comment (jamesdphillips) [inline api/core/v2/event_sort.go:49]: 👍

File: backend/store/store.go (hunk 1 above)
Comment (jamesdphillips) [reply backend/store/store.go:103]: never mind me

File: backend/store/store.go [hunk 3]
```diff
@@ -99,14 +99,18 @@ type SelectionPredicate struct {
...
 	// Limit indicates the number of resources to retrieve
 	Limit int64
 	// Subcollection represents a sub-collection of the primary collection
 	Subcollection string
+	// Ordering indicates the property to sort on, if supported by the store
+	Ordering string
```
Comment (ccressent) [inline backend/store/store.go:109]: Can you also add `OrderingDirection string` (which will hold one of the constants you've defined fore `ASC/DESC`)?

File: backend/store/store.go (hunk 1 above)
Comment (ccressent) [reply backend/store/store.go:103]: My take is that we shouldn't need this field so you can remove it.

Comment (jamesdphillips) [reply backend/store/store.go:103]: Unless I'm mistaken, my only concern is that the resolver would need to fabricate a continue token

**Review by jamesdphillips — APPROVED** [review]

File: api/core/v2/event_sort_test.go [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+			input:    []*Event{critical, noCheck, ok},
+			expected: []*Event{critical, ok, noCheck},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			sort.Sort(EventsBySeverity(tc.input, false))
+			fmt.Println("EXP:", eventsToId(tc.expected))
+			fmt.Println("RES:", eventsToId(tc.input))
```
Comment (echlebek) [inline api/core/v2/event_sort_test.go:245]: tests should not fmt.Println please :)

File: backend/store/store.go [hunk 5]
```diff
@@ -103,10 +103,14 @@ type SelectionPredicate struct {
...
+	// OrderDirection indicates the sorting direction
+	OrderingDirection string
```
Comment (echlebek) [inline backend/store/store.go:109]: Not to bikeshed too much here but this really only needs to be a boolean.

What about OrderingDirection bool with ASC and DESC constants for true/false, or a field named Descending which defaults to false by way of zero value?

File: api/core/v2/event_sort_test.go (hunk 4 above)
Comment (fguimond) [reply api/core/v2/event_sort_test.go]: Yes agreed, thanks for catching it!

File: backend/store/store.go (hunk 5 above)
Comment (fguimond) [reply backend/store/store.go:109]: I like the bool idea. It probably would be up to the store itself to convert it to whatever format it needs to.

Comment (fguimond) [reply backend/store/store.go]: Updated.

**Review by echlebek — APPROVED** [review]

**Pull request merged**